src/index.js

```diff
@@ -95,23 +71,28 @@ class ProcessManager {
    */
 
   hook(type, ...args) {
-    const hooks = _.filter(this.hooks, { type });
-    const promises = _.map(hooks, ({ handler }) => reflect(handler, args));
-
-    return Promise.all(promises)
-      .timeout(this.timeout, type)
-      .catch(Promise.TimeoutError, () => {
-        for (let i = 0; i < hooks.length; ++i) {
-          if (!promises[i].isPending()) {
-            continue;
-          }
-
+    const hooks = this.hooks.filter(hook => hook.type === type);
+    const promises = hooks.map(({ handler }) => utils.reflect(handler, args));
+
+    return Promise.race([
+      Promise.all(promises),
+      utils.delay(this.timeout)
+    ])
+    .then(async () => {
+      for (let i = 0; i < hooks.length; ++i) {
+        if (inspect(promises[i]).includes('<pending>')) {
           this.errors.push(hooks[i].timeoutError);
 
           log.info(`Timeout: ${hooks[i].name} took too long to complete ${type} hook`);
+        } else {
+          const error = await promises[i];
+
+          if (error) {
+            this.errors.push(error);
+          }
         }
-      })
-      .then(errors => this.errors.push(..._.compact(errors)));
+      }
+    });
```

Comment:
I'm not a fan of using util.inspect to check the status of a promise.
Can't we move the race inside the promises map and do a allSettled instead?

```js
  hook(type, ...args) {
    const hooks = this.hooks.filter(hook => hook.type === type);
    const promises = hooks.map(({ handler }) => Promise.race([utils.reflect(handler, args), utils.timeout(this.timeout)]));

    return Promise.allSettled(promises)
    .then(async () => {
      for (let i = 0; i < hooks.length; ++i) {
        if (promises[i].status === 'rejected') {
          this.errors.push(hooks[i].timeoutError);

          log.info(`Timeout: ${hooks[i].name} took too long to complete ${type} hook`);
        } else if (promises[i].value) {
          this.errors.push(promises[i].value);
        }
      }
    });
  }
```

And add the util:

```js
/**
 * Creates promise that will reject after the given time (in ms).
 */

module.exports.timeout = function(ms) {
  return new Promise((resolve, reject) => setTimeout(reject, ms));
};
```

Reply:
I was avoiding bumping the minimum node engine :smile:

Reply:
The CI should reflect that then.
Run a matrix with multiple node versions.

Promise.allSettled is node 12.9+, should be widespread by now.